### PR TITLE
Feature/add national insurance component

### PR DESCRIPTION
**Pull request by MCrawleyHomeOffice**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
Please also include any acceptance criteria if you have any.

- Added a new field for National Insurance Number to both the designer and runner
- Added appropriate tests in both designer and runner

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce
the testing if necessary.

Before PR's can be merged they will need to be tested by QA and approved where
applicable. To flag the change to QA assign **@XGovFormBuilder/qa** as one of the reviewers.

- [x] Test that you can add a National Insurance Field in the designer and then modify it
- [x] Test that the National Insurance Field is displayed in the Runner and that it validates user input as expected - Field should accept uppercase and lowercase, with and without spaces

# Checklist:

- [x] My changes do not introduce any new linting errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation and versioning
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have rebased onto main and there are no code conflicts
- [x] I have updated the architecture diagrams as per Contribute.md

### CODE REVIEW

File: designer/client/components/ComponentCreate/__tests__/ComponentCreateList.test.tsx [hunk 1]
```diff
@@ -14,7 +14,7 @@ const { beforeEach, suite, test, after } = lab;
 
 initI18n();
 
-suite("ComponentCreateList", () => {
+suite.only("ComponentCreateList", () => {
```
Comment (jenbutongit) [inline designer/client/components/ComponentCreate/__tests__/ComponentCreateList.test.tsx:17]: remove .only

File: designer/client/component.js [hunk 2]
```diff
@@ -45,6 +46,14 @@ function ComponentField(props) {
...
 
+function NationalInsuranceField() {
+  return (
+    <ComponentField>
+      <div className="box" />
+    </ComponentField>
+  );
+}
+
 function TextField() {
```
Comment (jenbutongit) [inline designer/client/component.js:57]: since these two components display the same thing, we can just remove `function NationalInsuranceField`, and in the componentTypes object, change `NationalInsuranceField: NationalInsuranceField` to `NationalInsuranceField: TextField`

File: runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        ["AA 12 34 56 C", true],
+        ["aa 12 34 56 c", true],
+        ["aa123456c", true],
+        ["", false],
+        [1234567890, false],
+        ["aa 12 34 56 c1", false],
+        ["aa123456c1", false],
+        ["aa123456", false],
+        ["AA123456C1", false],
+        ["AA123456", false],
```
Comment (jenbutongit) [inline runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts:60]: can we add a test for QQ123456C? we should see that it fails.

Comment (jenbutongit) [inline runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts:60]: it looks like you're testing NIs that are
- in caps with no whitespace passes
- in caps with whitespaces passes
- in lowercase with whitespace
- in lowercase without whitespace
- empty 
- invalid format
- too long lowercase
- too short
- too long in caps
- too short in caps 

IMO, there doesn't need to be this many tests. not that it affects run time, but just because it means there's more to maintain. 

- in caps without whitespace
- in caps with whitespaces 
- in lowercase with whitespace
- in lowercase without whitespace <- you already have covered that lowercase, caps, whitespace, no whitespace
- empty 
- invalid format
- too long lowercase 
- too short lowercase  
- too long in caps <- not required, since you already have a case for too long
- too short in caps <- not required, since you have already have a case for too short


You should also consider splitting these out to `it("validates and passes"` (or something to that nature), and `it("validates and fails")` - then you don't have to do the if statement.

File: runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        formModel
+      );
+
+      testCases.forEach((testCase: any) => {
+        const [nino, shouldPass] = testCase;
+        if (shouldPass) {
+          expect(component.formSchema.validate(nino).error).to.not.exist();
+        } else {
+          expect(component.formSchema.validate(nino).error).to.exist();
+        }
```
Comment (jenbutongit) [inline runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts:77]: I tend to avoid iterating in unit tests - usually because when these fail, they can be unhelpful messages. As in "error on line 74" - thanks, but what caused it! 

I would probably do inside `it("does not produce an error when the NI is valid")..` 
```.js
// init the component etc.. 
expect(component.formSchema.validate("AA123456C").error).to.not.exist();
expect(component.formSchema.validate("AA 12 34 56 C").error).to.not.exist();
//etc.. 
```

**Review by jenbutongit — CHANGES_REQUESTED** [review]
mostly LGTM! Just a few comments around the unit tests. 💥

File: runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts (hunk 4 above)
Comment (MCrawleyHomeOffice) [reply runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts:78]: Resolved.

File: runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts (hunk 3 above)
Comment (MCrawleyHomeOffice) [reply runner/test/cases/server/plugins/engine/nationalinsurancefield.test.ts:60]: Resolved.

File: designer/client/component.js (hunk 2 above)
Comment (MCrawleyHomeOffice) [reply designer/client/component.js:57]: Resolved.

File: designer/client/components/ComponentCreate/__tests__/ComponentCreateList.test.tsx (hunk 1 above)
Comment (MCrawleyHomeOffice) [reply designer/client/components/ComponentCreate/__tests__/ComponentCreateList.test.tsx]: Resolved.

**Review by jenbutongit — APPROVED** [review]

**Comment by MCrawleyHomeOffice** [comment]
As discussed with QA, we'd like to add additional regex validation to this to account for the rules documented here - https://www.gov.uk/hmrc-internal-manuals/national-insurance-manual/nim39110

File: designer/client/component.js [hunk 5]
```diff
@@ -35,6 +35,7 @@ export const componentTypes = {
   FlashCard: FlashCard,
   List: List,
   WarningText: WarningText,
+  NationalInsuranceField: TextField,
```
Comment (sulthan-ahmed) [inline designer/client/component.js:38]: just a learner question 
why is there a comma at the end of this property before the `}` within an object?
IntelliJ told me to put a comma too.  I thought this wasn't allowed in JS

Comment (jenbutongit) [reply designer/client/component.js:38]: we have eslint and prettier set up so that we add a dangling comma at the end of objects. It's been valid JS since ES5. dangling/trailing commas are useful for helping with merge conflicts. If you were to add another property at the end and there was no dangling comma, the diff would make it look like you changed  `NationalInsuranceField: TextField` - because you had to add a `,` so it's  
```
NationalInsuranceField: TextField, 
TheNextField: TextField
```
when actually, you didn't change that line at all. 

tldr: it helps with merge conflicts and/or diffs for the reviewer

File: designer/client/components/ComponentCreate/__tests__/__snapshots__/ComponentCreateList.jest.tsx.snap [hunk 6]
```diff
@@ -233,6 +233,19 @@ exports[`ComponentCreateList should match snapshot 1`] = `
               A text box for users to enter multiple lines of text into, for example as 'Additional info'
             </div>
           </li>
+          <li>
```
Comment (sulthan-ahmed) [inline designer/client/components/ComponentCreate/__tests__/__snapshots__/ComponentCreateList.jest.tsx.snap:236]: what is this file for?

File: designer/client/component.js (hunk 5 above)
Comment (sulthan-ahmed) [reply designer/client/component.js:38]: ooh useful to know, thanks :)

File: designer/client/components/ComponentCreate/__tests__/__snapshots__/ComponentCreateList.jest.tsx.snap (hunk 6 above)
Comment (jenbutongit) [reply designer/client/components/ComponentCreate/__tests__/__snapshots__/ComponentCreateList.jest.tsx.snap:236]: I generally advise against snapshots since they can end up being flakey/annoying to deal with/run slower but can have their uses

Comment (sulthan-ahmed) [reply designer/client/components/ComponentCreate/__tests__/__snapshots__/ComponentCreateList.jest.tsx.snap:236]: ok, thanks, I'll come back to this part after in my PR :)

**Review by jenbutongit — APPROVED** [review]
LGTM 🥇

**Comment by jenbutongit** [comment]
Closing to clear up our PR list :)
